### DISCUSSION TITLE
GBAK v5 - Restore from 2.5 - What are BLRs in restored DB will be recreated by GBAK?

### DISCUSSION THREAD

**Discussion by AnatolyLoy** [FirebirdSQL/firebird#8261] [Q&A] [answered]
**Context:** 
During database migration, we catch a case when the stored procedure only works correctly in Firebird 5.0 once recompiled (recreated or altered). We found that BLRs # 1 (after restoration) and # 2 (after recompilation) were differed.
As a solution, we decide to use recompilation of all procedures and triggers as part of the migration process.
However, database metadata also contains other types of BLRs for objects.
So, any BLR not recompiled during database restoration may differ from the BLR created by extracting metadata and creating an empty database.
In FirebirdSQL 5.0, we have the next BLRs:
- Procedures, functions and triggers
  - RDB$PROCEURES.RDB$PROCEDURE_BLR
  - RDB$FUNCTIONS.RDB$FUNCTION_BLR
  - RDB$TRIGGERS.RDB$TRIGGER_BLR
- Views
  - RDB$RELATIONS.RDB$VIEW_BLR
  - RDB$FIELDS.RDB$VALIDATE_BLR
- Domains and columns
  - RDB$FIELDS.RDB$VALIDATION_BLR
  - RDB$FIELDS.RDB$COMPUTED_BLR
- Indices
  - RDB$INDICES.RDB$EXPRESSION_BLR
  - RDB$INDICES.RDB$CONDITION_BLR

**Question:**
What BLRs are copying during restoration, and what are re-creating by GBAK 5.0?

**Comment by aafemt** [comment, marked as answer]
No BLR is recreated during restore. Everything is copied as is.

**Reply by AnatolyLoy** [reply]
So, any source associated with BLR and re-applied to a database restored by GBAK from 2.5 to 5.0 **may have** different BLR content and, therefore, different behaviour. Is it right?
So, after migration, we must recompile stored procedures, triggers, and views to ensure that they work the same as objects directly created in Firebird 5.0.

However, domains, columns, and indices expressions are elementary, and the possibility of a difference is low or even zero. Is that right? If the possibility is not zero, we need to alter such objects and the backup-restore database again to ensure they are robust and behave the same as directly created objects.

**Reply by aafemt** [reply]
Yes, BLR and behavior is changed from version to version. That's why migration by gbak is NOT the best way, only the simplest one.

**Reply by alexeykovyazin** [reply]
@AnatolyLoy Can you please share the metadata (from gbak -b -m)  of the original database and example of stored procedure call with differnt behaviour?

**Reply by asfernandes** [reply]
> Yes, BLR and behavior is changed from version to version. That's why migration by gbak is NOT the best way, only the simplest one.

This is incorrect answer. There is no such recommendation.

**Reply by aafemt** [reply]
There is no word "recommendation" in the quoted message. On purpose.

**Reply by AnatolyLoy** [reply]
> @AnatolyLoy Can you please share the metadata (from gbak -b -m) of the original database and example of stored procedure call with differnt behaviour?

@alexeykovyazin , I made the GDB-file and test script; see the attached archive.
[TEST__BLR25_50_BROKEN.zip](https://github.com/user-attachments/files/17085445/TEST__BLR25_50_BROKEN.zip)
